File: desktop/src/support/webcore/rules.cpp
#include "rules.hpp"

#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <algorithm>
#include <cmath>

namespace stencil::support {

  namespace {
    QStringList strings(const QJsonArray& a) {
      QStringList out;
      for (const QJsonValue& v : a) out << v.toString();
      return out;
    }

    QVector<QRect> rects(const QJsonArray& a) {
      QVector<QRect> out;
      for (const QJsonValue& v : a) {
        const QJsonArray r = v.toArray();
        if (r.size() == 4) out.push_back(QRect(r.at(0).toInt(), r.at(1).toInt(), r.at(2).toInt(), r.at(3).toInt()));
      }
      return out;
    }

    void readTokens(const QJsonObject& o, QHash<QString, QColor>& into) {
      for (auto it = o.begin(); it != o.end(); ++it) into.insert(it.key(), QColor(it.value().toString()));
    }

    WebcoreConfig parseConfig() {
      WebcoreConfig c;
      QFile f(":/config/webcore.json");
      if (!f.open(QIODevice::ReadOnly)) return c;
      const QJsonObject root = QJsonDocument::fromJson(f.readAll()).object();
      const QJsonObject tokens = root.value("tokens").toObject();
      readTokens(tokens.value("light").toObject(), c.tokens[0]);
      readTokens(tokens.value("dark").toObject(), c.tokens[1]);
      const QJsonObject font = root.value("font").toObject();
      c.fontFamilies = strings(font.value("families").toArray());
      c.fontPx = font.value("px").toInt(c.fontPx);
      c.fontPt = font.value("pt").toInt(c.fontPt);
      const QJsonObject img = root.value("image").toObject();
      c.width = img.value("width").toInt(c.width);
      c.height = img.value("height").toInt(c.height);
      c.cell = img.value("cell").toInt(c.cell);
      c.horizonShare = img.value("horizonShare").toDouble(c.horizonShare);
      c.skyBands = strings(img.value("skyBands").toArray());
      const QJsonObject hill = img.value("hill").toObject();
      c.ampCells = hill.value("ampCells").toInt(c.ampCells);
      c.spanCells = hill.value("spanCells").toInt(c.spanCells);
      c.offsetCells = hill.value("offsetCells").toInt(c.offsetCells);
      c.grass = strings(img.value("grass").toArray());
      c.cloudInk = img.value("cloudInk").toString();
      c.cloudShade = img.value("cloudShade").toString();
      for (const QJsonValue& v : img.value("clouds").toArray()) {
        const QJsonObject o = v.toObject();
        WebcoreCloud cloud;
        cloud.x = o.value("x").toInt();
        cloud.y = o.value("y").toInt();
        cloud.blocks = rects(o.value("blocks").toArray());
        cloud.shade = rects(o.value("shade").toArray());
        c.clouds.push_back(cloud);
      }
      const QJsonObject word = root.value("word").toObject();
      c.text = word.value("text").toString();
      c.stroke = word.value("stroke").toString();
      c.thickness = word.value("thickness").toDouble(c.thickness);
      c.colors = strings(word.value("colors").toArray());
      const QJsonArray grid = word.value("grid").toArray();
      if (grid.size() == 2) { c.gridW = grid.at(0).toInt(); c.gridH = grid.at(1).toInt(); }
      c.gapCells = word.value("gapCells").toInt(c.gapCells);
      c.widthShare = word.value("widthShare").toDouble(c.widthShare);
      c.centerYShare = word.value("centerYShare").toDouble(c.centerYShare);
      const QJsonObject glyphs = word.value("glyphs").toObject();
      for (auto it = glyphs.begin(); it != glyphs.end(); ++it) {
        QVector<QPoint> poly;
        for (const QJsonValue& p : it.value().toArray()) {
          const QJsonArray xy = p.toArray();
          if (xy.size() == 2) poly.push_back(QPoint(xy.at(0).toInt(), xy.at(1).toInt()));
        }
        if (!it.key().isEmpty()) c.glyphs.insert(it.key().at(0), poly);
      }
      const QJsonObject s = root.value("strings").toObject();
      c.offToast = s.value("off").toString();
      c.projectName = s.value("project").toString();
      c.imageName = s.value("imageName").toString();
      return c;
    }

    double round2(double v) { return std::round(v * 100) / 100; }
  }  // namespace

  const WebcoreConfig& webcoreConfig() {
    static const WebcoreConfig cfg = parseConfig();
    return cfg;
  }

  CellGrid cellGrid() {
    const WebcoreConfig& c = webcoreConfig();
    CellGrid g;
    g.cell = c.cell;
    g.cols = c.width / c.cell;
    g.rows = c.height / c.cell;
    g.horizon = qRound(g.rows * c.horizonShare);
    return g;
  }

  int hillTopCell(int cx, const CellGrid& grid) {
    const WebcoreConfig& c = webcoreConfig();
    const int u = (((cx - c.offsetCells) % c.spanCells) + c.spanCells) % c.spanCells;
    const double lift = std::floor((double(c.ampCells) * 4 * u * (c.spanCells - u)) / (double(c.spanCells) * c.spanCells));
    return grid.horizon - int(lift);
  }

  QString skyBandAt(int cy, const CellGrid& grid) {
    const QStringList& bands = webcoreConfig().skyBands;
    const int i = std::min(int(bands.size()) - 1, (cy * int(bands.size())) / grid.horizon);
    return bands.value(i);
  }

  QString grassShadeAt(int cx, int cy, const CellGrid& grid) {
    const WebcoreConfig& c = webcoreConfig();
    const int bands = int(c.grass.size());
    const int bandCells = std::max(1, (grid.rows - grid.horizon + c.ampCells) / bands);
    const int d = cy - hillTopCell(cx, grid) + ((cx + cy) & 1) * (bandCells >> 1);
    return c.grass.value(std::min(bands - 1, d / bandCells));
  }

  QString cellColorAt(int cx, int cy, const CellGrid& grid) {
    return cy >= hillTopCell(cx, grid) ? grassShadeAt(cx, cy, grid) : skyBandAt(cy, grid);
  }

  QVector<CellRun> cloudBlocks() {
    const WebcoreConfig& c = webcoreConfig();
    QVector<CellRun> out;
    for (const WebcoreCloud& cl : c.clouds) {
      for (const QRect& r : cl.blocks) out.push_back({cl.x + r.x(), cl.y + r.y(), r.width(), r.height(), c.cloudInk});
      for (const QRect& r : cl.shade) out.push_back({cl.x + r.x(), cl.y + r.y(), r.width(), r.height(), c.cloudShade});
    }
    return out;
  }

  QVector<CellRun> cellRuns(const CellGrid& grid) {
    QVector<CellRun> runs;
    for (int cy = 0; cy < grid.rows; ++cy) {
      int start = 0;
      QString color = cellColorAt(0, cy, grid);
      for (int cx = 1; cx <= grid.cols; ++cx) {
        const QString c = cx < grid.cols ? cellColorAt(cx, cy, grid) : QString();
        if (c == color) continue;
        runs.push_back({start, cy, cx - start, 1, color});
        start = cx;
        color = c;
      }
    }
    return runs;
  }

  QString pixelColorAt(int px, int py) {
    const CellGrid grid = cellGrid();
    const int cx = px / grid.cell, cy = py / grid.cell;
    QString color = cellColorAt(cx, cy, grid);
    for (const CellRun& b : cloudBlocks())
      if (cx >= b.x && cx < b.x + b.w && cy >= b.y && cy < b.y + b.h) color = b.color;
    return color;
  }

  QVector<WordLine> wordLines(int w, int h) {
    const WebcoreConfig& c = webcoreConfig();
    const int n = int(c.text.size());
    const int totalCells = n * c.gridW + (n - 1) * c.gapCells;
    const double cellPx = (w * c.widthShare) / totalCells;
    const double left = (w - totalCells * cellPx) / 2;
    const double top = h * c.centerYShare - (c.gridH * cellPx) / 2;
    QVector<WordLine> out;
    for (int i = 0; i < n; ++i) {
      WordLine line;
      for (const QPoint& p : c.glyphs.value(c.text.at(i)))
        line.points.push_back(QPointF(round2(left + (i * (c.gridW + c.gapCells) + p.x()) * cellPx),
                                      round2(top + p.y() * cellPx)));
      line.color = c.stroke;
      line.fillColor = c.colors.value(i % int(c.colors.size()));
      line.thickness = c.thickness;
      out.push_back(line);
    }
    return out;
  }

}  // namespace stencil::support

File: browser/js/ui/webcore/rules.js
// The webcore skin's table (config/webcore.json) and the pure rules over it: the picture's
// cells and the word's lines. DOM-free; desktop twin: support/webcore/rules.{hpp,cpp}, value
// for value.
import WEBCORE_DATA from '../../config/webcore.json' with { type: 'json' };

export const WEBCORE = WEBCORE_DATA;
export const SKIN_ATTR = 'data-skin';
export const SKIN_NAME = 'webcore';
export const OFF_TOAST = WEBCORE.strings.off;
export const PROJECT_NAME = WEBCORE.strings.project;
export const IMAGE_NAME = WEBCORE.strings.imageName;

// The picture's cell grid: `horizon` is the row the hill's foot rests on.
export const cellGrid = () => {
  const { width, height, cell, horizonShare } = WEBCORE.image;
  const rows = Math.floor(height / cell);
  return { cell, cols: Math.floor(width / cell), rows, horizon: Math.round(rows * horizonShare) };
};

// The crest at column cx: a repeating parabola in integer arithmetic, so both painters agree.
export const hillTopCell = (cx, grid) => {
  const { ampCells, spanCells, offsetCells } = WEBCORE.image.hill;
  const u = (((cx - offsetCells) % spanCells) + spanCells) % spanCells;
  return grid.horizon - Math.floor((ampCells * 4 * u * (spanCells - u)) / (spanCells * spanCells));
};

export const skyBandAt = (cy, grid) => {
  const bands = WEBCORE.image.skyBands;
  return bands[Math.min(bands.length - 1, Math.floor((cy * bands.length) / grid.horizon))];
};

// Crest to foot in bands, a checker dither half a band deep at every edge.
export const grassShadeAt = (cx, cy, grid) => {
  const { grass, hill } = WEBCORE.image;
  const bandCells = Math.max(1, Math.floor((grid.rows - grid.horizon + hill.ampCells) / grass.length));
  const d = cy - hillTopCell(cx, grid) + ((cx + cy) & 1) * (bandCells >> 1);
  return grass[Math.min(grass.length - 1, Math.floor(d / bandCells))];
};

export const cellColorAt = (cx, cy, grid) =>
  (cy >= hillTopCell(cx, grid) ? grassShadeAt(cx, cy, grid) : skyBandAt(cy, grid));

// Every cloud block as {x, y, w, h, color} in cells: the ink first, the shade over it.
export const cloudBlocks = () => {
  const { clouds, cloudInk, cloudShade } = WEBCORE.image;
  const out = [];
  for (const c of clouds) {
    for (const [dx, dy, w, h] of c.blocks) out.push({ x: c.x + dx, y: c.y + dy, w, h, color: cloudInk });
    for (const [dx, dy, w, h] of c.shade || []) out.push({ x: c.x + dx, y: c.y + dy, w, h, color: cloudShade });
  }
  return out;
};

// Same-colour runs along each row: fewer rects, the same pixels.
export const cellRuns = (grid) => {
  const runs = [];
  for (let cy = 0; cy < grid.rows; cy++) {
    let start = 0, color = cellColorAt(0, cy, grid);
    for (let cx = 1; cx <= grid.cols; cx++) {
      const c = cx < grid.cols ? cellColorAt(cx, cy, grid) : null;
      if (c === color) continue;
      runs.push({ x: start, y: cy, w: cx - start, h: 1, color });
      start = cx;
      color = c;
    }
  }
  return runs;
};

// The colour the finished picture holds at pixel (px, py): the last cloud block over the cell.
export const pixelColorAt = (px, py) => {
  const grid = cellGrid();
  const cx = Math.floor(px / grid.cell), cy = Math.floor(py / grid.cell);
  let color = cellColorAt(cx, cy, grid);
  for (const b of cloudBlocks())
    if (cx >= b.x && cx < b.x + b.w && cy >= b.y && cy < b.y + b.h) color = b.color;
  return color;
};

const round2 = (v) => Math.round(v * 100) / 100;

// The word as closed, locked, filled lines across the sky of a w×h image, one colour each.
export const wordLines = (w, h) => {
  const { text, glyphs, grid, gapCells, widthShare, centerYShare, colors, stroke, thickness } = WEBCORE.word;
  const [gw, gh] = grid;
  const n = text.length;
  const totalCells = n * gw + (n - 1) * gapCells;
  const cellPx = (w * widthShare) / totalCells;
  const left = (w - totalCells * cellPx) / 2;
  const top = h * centerYShare - (gh * cellPx) / 2;
  return [...text].map((ch, i) => ({
    points: glyphs[ch].map(([x, y]) => ({
      x: round2(left + (i * (gw + gapCells) + x) * cellPx),
      y: round2(top + y * cellPx),
    })),
    locked: true,
    color: stroke,
    fillColor: colors[i % colors.length],
    thickness,
  }));
};
